src/Illuminate/Redis/Lua/LuaScript.php

```diff
@@ -0,0 +1,237 @@
+     *
+     * @return string|null
+     */
+    public function getSha1()
+    {
+        return $this->sha1;
+    }
+
+    /**
+     * Set the list of keys and arguments to be passed to the script.
+     *
+     * @param  array  $keys  The list of keys to be passed to the script.
+     * @param  array|null  $arguments  The list of arguments to be passed to the script.
+     * @return $this
+     */
+    public function with($keys, $arguments = null)
+    {
+        $this->keys = $keys;
+        $this->arguments = $arguments;
+
+        return $this;
+    }
+
+    /**
+     * Execute the Lua script with the given SHA1 hash using the given Redis connection.
+     *
+     * @param  string  $sha  The SHA1 hash of the Lua script.
+     * @return mixed The result of the script execution.
+     */
+    protected function executeWithHash($sha)
+    {
+        return $this->connection->command('evalsh', [$sha, $this->keys, $this->arguments, count($this->keys)]);
+    }
+
+    /**
+     * Execute the Lua script using EVAL command with the given Redis connection.
+     *
+     * @return mixed The result of the script execution.
+     */
+    protected function executeWithScript()
+    {
+        return $this->connection->eval($this->script, count($this->keys), $this->keys, $this->arguments);
+    }
+
+    /**
+     * Execute the Lua script using the given Redis connection.
+     *
+     * @return mixed The result of the script execution.
+     *
+     * @throws \Illuminate\Redis\Lua\Exception\LuaScriptLoadException If the lua script could not be loaded.
+     */
+    public function execute()
+    {
+        if ($this->sha1 !== null) {
+            return $this->executeWithHash($this->sha1);
+        } else {
+            if ($this->isCachingEnable) {
+                // Try to execute the script.
+                $hash = sha1($this->script);
+                $result = $this->executeWithHash($hash);
+
+                if ($result !== false) {
+                    return $result;
+                }
+
+                // Maybe script not exist in redis, Try to load that
+                if ($this->loadScript() === false) {
+                    throw new LuaScriptLoadException;
+                }
+
+                return $this->executeWithHash($hash);
+            } else {
+                return $this->executeWithScript();
+            }
+        }
+    }
+
+    /**
+     * Load this lua script to redis.
+     *
+     * @return false|string Returns the SHA1 hash of the loaded script on success, or false on failure.
+     */
+    public function loadScript(): false|string
+    {
+        return $this->connection->script('load', $this->script);
+    }
+
+    /**
+     * Check is provided SHA1 key is exists in Redis server.
+     *
+     * @return bool A boolean indicating whether the provided SHA1 key exists in the Redis server.
+     */
+    public function exist(): bool
+    {
+        if ($this->sha1 === null) {
+            return true;
```

Comment:
Why is true? If it some kind of workaround for plain script support. it isn't intuitive.

Reply:
Possibly can't be called a workaround for plain script support, The `exists` method is only useable when we already provide a script sha1 hash, When we have a plain script it should return `true` or `null` or maybe throw some exception.

If you had a better idea to handle this I happy to know that

Reply:
Is "not provided sha1" "exists" in Redis? I think:
- no (false).
- error (null or exception).
null is also "falsy" value, so both negative cases can be checked with a single if.

Next imagine this situation. We get sha1 from some source. Can we call it? Lets check it first?
```
$sha1 = Cache::get('key_for_sha1');
$luaScript = LuaScript::fromSha1(Redis::connection(), $sha1);
if ($luaScript->exists()) {
    $luaScript->execute();
}
```
With a 'true' approach we will fail, because it's never been tested for sha1 correctness. We have to make extra check before instantiating LuaScript from it. Extra check is OK. But we either can throw `InvalidArgumentException` from `fromSha1()` for null to force the error (and to force developer to write extra check), before our code goes to far.

Second thought on it is what about caching? Will we cache our script it would become available via hash. So, asking for Redis "Is this script exists (loaded)" should return true - you're right. But until we cache this script it should return false. Perhaps we need to fill sha1 field after successful call when caching is enabled.

So what the essence of the `exists` method?
- Is the provided sha1 exists in redis? It shouldn't return true for sha1 = null. Will redis itself return true?
- Does the script can be called? It shouldn't return true, if we provide neither sha1 nor script (this check is missed).